src/lib/x509/certstor_flatfile/certstor_flatfile.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+* Certificate Store
+* (C) 1999-2019 Jack Lloyd
+* (C) 2019      Patrick Schmidt
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/build.h>
+
+#include <botan/certstor_flatfile.h>
+#include <botan/data_src.h>
+#include <botan/pem.h>
+
+namespace Botan {
+namespace {
+std::vector<secure_vector<uint8_t>> decode_all(DataSource& source)
+   {
+   std::vector<secure_vector<uint8_t>> pems;
+
+   while(!source.end_of_data())
+      {
+      std::string label;
+      try
+         {
+         pems.push_back(PEM_Code::decode(source, label));
+         }
+      catch(const Decoding_Error&) {}
+      }
+
+   return pems;
+   }
+}
+
+Flatfile_Certificate_Store::Flatfile_Certificate_Store()
+   {
+   }
+
+Flatfile_Certificate_Store::Flatfile_Certificate_Store(const std::string& file)
+   {
+   if(file.empty())
+      {
+      return;
+      }
+
+   DataSource_Stream file_stream(file);
+
+   for(const secure_vector<uint8_t> der : decode_all(file_stream))
+      {
+      std::shared_ptr<const X509_Certificate> cert = std::make_shared<const X509_Certificate>(der.data(), der.size());
+
+      m_all_subjects.push_back(cert->subject_dn());
+      m_dn_to_cert.emplace(cert->subject_dn(), cert);
+      m_pubkey_sha1_to_cert.emplace(cert->subject_public_key_bitstring_sha1(), cert);
+      m_subject_dn_sha256_to_cert.emplace(cert->raw_subject_dn_sha256(), cert);
+      }
+   }
```

Comment:
I think there should be a check that we successfully decoded at least one certificate from the file. Otherwise misconfiguration (say pointing to some file with no PEMs) results in silent failure and then a likely very confusing error when validation occurs about no trust root located. Otoh exactly 1 cert makes perfect sense when pinning.

Also check that all the certificates loaded are self signed CA certs - cert.is_self_signed() && cert.is_CA_cert() since any other certificate being in the trust bundle is probably an error.

Reply:
> Also check that all the certificates loaded are self signed CA certs - cert.is_self_signed() && cert.is_CA_cert() since any other certificate being in the trust bundle is probably an error.

 
This fails on my manjaro machine because `VeriSign Class 3 Public Primary Certification Authority - G3` isn't a CA.

Reply:
Ah this is a v1 certificate. IIRC v1 self-signed is supposed to be implicitly CA. So this is a bug on our part.

Reply:
Fixed in #1890

Reply:
I rebased to latest master. Works. Thanks.